lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscovery/RandomSpec.hs

```diff
@@ -0,0 +1,132 @@
+    , XPrv
+    , fromMnemonic
+    )
+import Cardano.Wallet.Primitive.AddressDerivation.Common
+    ( Key (..) )
+import Cardano.Wallet.Primitive.AddressDerivation.Random
+    ( deriveAccountPrivateKey
+    , deriveAddressPrivateKey
+    , generateKeyFromSeed
+    , minSeedLengthBytes
+    , toAddress
+    )
+import Cardano.Wallet.Primitive.AddressDerivationSpec
+    ( DecodeDerivationPath (..)
+    , decodeTest1
+    , decodeTest2
+    , decodeTest3
+    , decodeTest4
+    )
+import Cardano.Wallet.Primitive.AddressDiscovery
+    ( IsOurs (..) )
+import Cardano.Wallet.Primitive.AddressDiscovery.Random
+    ( RndState (..) )
+import Cardano.Wallet.Primitive.Types
+    ( Address (..) )
+import Test.Hspec
+    ( Expectation, Spec, describe, it, shouldBe )
+import Test.QuickCheck
+    ( Arbitrary (..), Property, choose, property, vectorOf, (.&&.), (===) )
+
+import qualified Data.ByteArray as BA
+import qualified Data.ByteString as BS
+
+spec :: Spec
+spec = do
+    describe "Random Address Discovery Properties" $ do
+
+        it "isOurs works as expected during key derivation" $ do
+            property prop_keyDerivationObeysIsOurs
+
+    goldenSpec
+
+
+{-------------------------------------------------------------------------------
+                                  Golden tests
+-------------------------------------------------------------------------------}
+
+goldenSpec :: Spec
+goldenSpec = describe "Golden tests" $ do
+
+    it "check isOurs - mainnet - initial account" $
+        checkIsOurs decodeTest1
+
+    it "check isOurs - mainnet - another account" $
+        checkIsOurs decodeTest2
+
+    it "check isOurs - testnet - initial account" $
+        checkIsOurs decodeTest3
+
+    it "check isOurs - testnet - another account" $
+        checkIsOurs decodeTest4
+
+checkIsOurs :: DecodeDerivationPath -> Expectation
+checkIsOurs DecodeDerivationPath{..} =
+    isOurs (Address addr) (RndState rootXPrv) `shouldBe` (True, RndState rootXPrv)
+  where
+    rootXPrv = generateKeyFromSeed (Passphrase seed) (Passphrase "")
+    Right (Passphrase seed) = fromMnemonic @'[12] mnem
+
+
+{-------------------------------------------------------------------------------
+                               Properties
+-------------------------------------------------------------------------------}
+
+prop_keyDerivationObeysIsOurs
+    :: Passphrase "seed"
+    -> Passphrase "encryption"
+    -> Index 'Hardened 'AccountK
+    -> Index 'Soft 'AddressK
+    -> Key 'RootK XPrv
+    -> Property
+prop_keyDerivationObeysIsOurs seed encPwd accIx addrIx rk' =
+    isOurs address (RndState rootXPrv) === (True, RndState rootXPrv) .&&.
+    isOurs address (RndState rk') === (False, RndState rk')
+  where
+    rootXPrv@(Key rootXPrv') = generateKeyFromSeed seed encPwd
+    rootXPub = Key $ toXPub rootXPrv'
+    accXPrv = deriveAccountPrivateKey encPwd rootXPrv accIx
+    (Key addrXPrv) = deriveAddressPrivateKey encPwd accXPrv addrIx
+    addrXPub = Key $ toXPub addrXPrv
```

Comment:
Minor here but: this should use `toPublicKey` to transform a `Key _ XPrv` and not break the abstraction boundary.